src/coreclr/vm/jitinterface.cpp

```diff
@@ -649,6 +649,44 @@ int CEEInfo::getStringLiteral (
     return result;
 }
 
+bool CEEInfo::tryGetNonRandomizedHashCode (
+        CORINFO_MODULE_HANDLE       moduleHnd,
+        mdToken                     metaTOK,
+        int*                        pHashCode)
+{
+    CONTRACTL{
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    bool result = false;
+
+    JIT_TO_EE_TRANSITION();
+
+    if (!IsDynamicScope(moduleHnd))
+    {
+        // If ResolveStringRef returns a pinned reference we can return it by value (IAT_VALUE)
+        CORINFO_OBJECT_HANDLE pPinnedString = nullptr;
+        reinterpret_cast<Module*>(moduleHnd)->ResolveStringRef(metaTOK, reinterpret_cast<void**>(&pPinnedString));
+        if (pPinnedString != nullptr)
+        {
+            GCX_COOP();
+            PREPARE_NONVIRTUAL_CALLSITE(METHOD__STRING__GETNONRANDOMIZEDHASHCODE);
```

Comment:
Invoking CoreLib at runtime won't work for AOT. What's your plan there?

Reply:
just include & use the managed impl directly? (probably by moving that method to a separate cs file)

Reply:
> It is almost true for current implementation of GetNonRandomizedHashCode, except of endianness.

In an unlikely event of supporting BE in R2R/NAOT, we can just limit the opt for Host being LE and target being LE for simplicity?

> It would break further if the implementation is changed e.g. to hash native int at a time on 64-bit architectures.

Same here - limit to x64 on x64? Also, native-int might not be a good idea due to 4-byte alignment of string data.

If NAOT/R2R support raises concerns we can probably leave it to be JIT only?

Reply:
You can always make the shared file work for NAOT/R2R with enough if checks and ifdefs. My point is that it is de-facto a second implementation once you do that.

> we can just limit the opt for Host being LE and target being LE for simplicity

Nit: Shortcuts like this violate invariants that we try to maintain for cross-compilation. We want our cross-compilers to produce byte-to-byte identical output for given target, irrespective of the host that the compiler is running on.

> NAOT/R2R support raises concerns we can probably leave it to be JIT only?

Every time folks bring up complicated NAOT specific JIT optimizations, I am pointing out that it would be great if we just make the JIT optimizations work for R2R/AOT - we are not there today that's unfortunate.

Reply:
> Every time folks bring up complicated NAOT specific JIT optimizations, I am pointing out that it would be great if we just make the JIT optimizations work for R2R/AOT where possible - we are not there today that's unfortunate.

I share those views, I only meant that if NAOT impl is the main concern - should we give up completely on some theoretical optimization even if the mainstream runtime can benefit from it just fine? (of course, it's possible to break this specific one too by introducing implicit side effects in the implementation).

Reply:
We should go extra mile to properly enable optimizations for AOT where possible. I do not have a problem with having JIT specific optimizations that are fundamentally incompatible with AOT (not the case here).

Reply:
So if I read this all correctly - NAOT impl can't be done without introducing coupling. We can't land without NAOT. the extra coupling needs real world justification. To be honest, I don't have any specific example, I've vibe-coded a simple analyzer that tries to catch lookups (Constains, TryGetValue) with literals and named constants as keys - https://gist.github.com/EgorBo/e3ee6c711741956b423fb0ccdd0b51c4

Roslyn: 84 matches
runtime: 54 matches
aspnetcore: 25 matches
OrchardCore: 54 matches

(ignoring folders like 'tests' etc.)

I guess I'll put this on pause till some use-case appears.

Reply:
Are there any in your list that look hot?

> We can't land without NAOT. 

I would not say that. I can accept it with a good reason.

Reply:
I totally forgot to mention that on NativeAOT it can only work with proper static PGO data, which is, presumably, never supplied on practice. Because this relies on:
1) FindValue is expected to be inlined (on JIT Dynamic PGO boosts it)
2) _comparer must be devirtualized under a guard (GDV)